File: python/Q1_two_sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#


# @lc code=start
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        mapping: dict[int, int] = dict()

        for i, n in enumerate(nums):
            if n in mapping:
                return [i, mapping[n]]

            mapping[target - n] = i

        return []


# @lc code=end

File: rust/src/leetcode/Q1_two_sum.rs
/*
 * @lc app=leetcode id=1 lang=rust
 *
 * [1] Two Sum
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut result: HashMap<i32, i32> = HashMap::new();

        for (i, &num) in nums.iter().enumerate() {
            if let Some(&idx) = result.get(&num) {
                return vec![idx, i as i32];
            }

            result.insert(target - num, i as i32);
        }

        vec![]
    }
}
// @lc code=end
